### ISSUE TITLE
'Toggle' and similar methods for lists

### ISSUE THREAD

**Issue by andybak** [matiboy/django-admin-methods#4]
I quite often implement 'action' columns in changelists. For example an app might have a column for 'archive' that sets a flag on the model and changes to 'unarchive' if it's already set.

This is fairly close to your 'toggle' action except part of the list.

To avoid the need to modify urls.py we can override ModelAdmin.get_urls similar to this: https://github.com/crccheck/django-object-actions/blob/master/django_object_actions/utils.py#L35 (this does require a mixin for your ModelAdmin but it could be optional)

Do you think this is a useful addition? If so I'll put something together.

**Comment by matiboy** [comment]
It's basically the same function as `toggle` but displayed on each row?
So that would belong in the `model` module I guess
Yes sounds useful!

**Comment by andybak** [comment]
On a side note - your project inspired me to clean up and extract some code I wrote for another project: https://github.com/DjangoAdminHackers/django-admin-row-actions

See what you think.

**Comment by matiboy** [comment]
Nice!

Do you think there would be a way to use normal fields instead of Json config? Kind of what Django admin does for single model views: 

``` python
  fields = ('last_name', ('first_name', 'middle_name'), 'photo')
```

where first_name and middle_name are typical admin fields and they get grouped into a single row.

**Comment by andybak** [comment]
I'm not sure as you need as a bare minimum for each menu item a label and a url. So it would have to be at least a list of tuples - and there's not much advantage to supporting that in addition to a dict.

If I were to take this further then allowing a class rather than a dict would be the way to go. Kind of like forms and inlines.

```
class FooRowActions(RowAction):
    print_pdf = row_actions.UrlAction(url='blah', parameters=('id', 'title')
    send_invoice = row_actions.CallableAction(fn='obj.send_invoice')
    ...

class FooAdmin(admin.ModelAdmin):
    row_actions = FooRowActions
```

or something like that.

**Comment by matiboy** [comment]
What if you were to use something similar to what I did on PR #7 whereby you can pass a callable attribute as the model value to be rendered. Then the label+url could be defined within that, which means you can reuse that model field function somewhere else?

For example:
admin.py

``` python
  class UnitAdmin(admin.ModelAdmin):
    unit_tenant = admin_methods.list.list('tenants', model_attribute='qr_code')

    list_display = ('name', 'group_count', 'unit_tenant',)

admin.site.register(Unit, UnitAdmin)
```

tenant model:

``` python
  class Tenant(django.db.models.Model):
    def qr_code(self):
        return django.utils.html.format_html('<img src="http://chart.apis.google.com/chart?cht=qr&chs=240x240&chl={}" />', self.token)

    qr_code.allow_tags = True
```

outputs

![image](https://cloud.githubusercontent.com/assets/487758/9325946/c967e4ae-45c8-11e5-89b6-6aa20ed5ebf6.png)

Would that allow to create the various label+url with just an attribute?

**Comment by matiboy** [comment]
@andybak Can you have a look at PR #8 ?
Let me know what you think.

It's a first draft

**Comment by andybak** [comment]
Looks good. One comment. I've always set the th caption to '' - the idea being that the action label in each row should be self-descriptive. Not sure if this is a better default than 'toggle x'.

I know it's only a tiny snippet but if the html was stored in a template then people could easily override it (i.e. a button instead of a link etc).

**Comment by matiboy** [comment]
Regarding the th header, having not used this yet, I'll follow your suggestion.

Great idea on the template, also allows to rely on the template rendering engine to decide whether the url part is a callable, and to use data from the instance which they could not have in description_true, description_false.

Code updated.

Thanks

**Comment by andybak** [comment]
Glad you found my suggestions useful.

One last thought about the templates (and this might be being over-fussy). It should be possible to pass in a template name to override loading the default - otherwise a user has to rely on INSTALLED_APPS ordering to ensure his will override the built-in one of the same name.

**Comment by matiboy** [comment]
Was just about to do that :laughing:

**Comment by matiboy** [comment]
Closed with PR #8
